src/jit/compiler.hpp

```diff
@@ -2708,6 +2708,26 @@ bool                    Compiler::compCanEncodePtrArgCntMax()
 #endif
 }
 
+// fgAreNextNodesValidForTailCall - Returns true if the fgMorphStmt contains the right pattern that 
+//                                  the tail call optimization expects.
+//
+// Arguments:
+//    fgMorphStmt      - The statement that contains the call expression to be converted to a tail call.
+inline
+bool Compiler::fgAreNextNodesValidForTailCall(GenTreePtr fgMorphStmt)
+{
+    GenTreePtr nextStmtExpr = fgMorphStmt->gtNext->gtStmt.gtStmtExpr;
+    return (nextStmtExpr->gtOper == GT_RETURN ||
+            // There could be a GT_COMMA generated between the call and the return.
+            // This happens when the result of the call is not used and discarded (in IL) by a pop.
+            // The generation of the pop IL instruction happens in "case CEE_POP:" in importer.cpp. The GT_COMMA
+            // is generated by the  gtUnusedValNode(op1) call and is intended to help the optimizers not to 
+            // change the call and its result in case of the call having side effects or the compilation happens
+            // for Debug configuration. Handle the potential existance of comma by checking that its second 
+            // operand is NOP and the next statement is a GT_RETURN.
+                (nextStmtExpr->gtOper == GT_COMMA && nextStmtExpr->gtOp.gtOp2->gtOper == GT_NOP &&
```

Comment:
@LLITCHEV explained offline that what is on the top of the stack is var=call, that is a GT_ASG node and hence CEE_POP results in adding a GT_COMMA node.

From importer.cpp code read, importer introduces temp assignments while generating debuggable code by calling impSpillStackEnsure().

when the top of the stack is a GT_ASG, CEE_POP don't need to generate a GT_COMMA node since Gt_ASG is not an expression.  The right fix is to modify CEE_POP if-condition as follows

```
if (op1->gtOper != GT_CALL && op1->gtOper != GT_ASG)
{
    op1 = `gtUnusedValNode(op1);
}
```